pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -360,3 +363,246 @@ def matrix_multiply_parallel(matrix_1, matrix_2, num_threads):
+
+    Parameters:
+     process_objects
+        accepts array of objects on which the algorithm is supposed to run
+     maximize_prob 
+        pass True if the algorithm should find maximum value of the cost function otherwise pass False
+     min_compare_len 
+        a positive number decides to which level of gap the algorithm can maintain while iterating from start to end,
+        for example-> if minimun length is 2 then it can only iterate if endIndex=startIndex+2
+     lookup_index 
+        format-->[start_index,endIndex] algorithm runs from start to end
+     get_lookup_fn
+      should return next range of indices
+      sample -> get_lookup_fn(position, rangeIndices, currentIndex)
+       position is either 'before' or 'after' 
+       rangeIndices is the present range of index like [start_index,endIndex]
+     cost_fn
+      should return the cost 
+      sample -> cost_fn(process_objects,rangeIndices,currentIndex)
+
+
+      **Usage examples : 
+
+      1.OPTIMAL BINARY SEARCH TREE
+
+        from binarytree import Node
+        n = 5
+        p = [None, Node(0.15), Node(0.10), Node(0.05), Node(0.10), Node(0.20)]
+        q = [Node(0.05), Node(0.10), Node(0.05), Node(0.05), Node(0.05), Node(0.10)]
+
+
+        def lookup(position, endIndex, middle):
+            if position is 'before':
+             return [endIndex[0], middle-1]
+            else:
+             return [middle+1, endIndex[1]]
+
+
+        def cost(obj, endIndex, middle):
+
+            if(endIndex[1]<endIndex[0]):
+                return obj['q'][endIndex[1]].value
+
+            sum = 0
+            for i in range(endIndex[0], endIndex[1]+1):
+                sum += obj['p'][i].value
+            for i in range(endIndex[0]-1, endIndex[1]+1):
+                sum += obj['q'][i].value
+            return sum
+
+
+        print(optimal_grouping({'p': p, 'q': q},  False, 1, [1, n], lookup, cost))
+
+
+
+      2.MATRIX CHAIN MULTIPLICATION
+
+        def cost(matrix, endIndex, middle):
+
+            if endIndex[0] == endIndex[1]:
+            return 0
+        return matrix[endIndex[0]-1]*matrix[middle]*matrix[endIndex[1]]
+
+
+        def lookup(position, endIndex, middle):
+        if position is 'before':
+            return [endIndex[0], middle]
+        else:
+            return [middle+1, endIndex[1]]
+
+
+        print(optimal_grouping([30, 35, 15, 5, 10, 20, 25], False, 2, [1, 6], lookup, cost))
+
+    """
+
+    if min_compare_len < 1:
+        raise ValueError(
+            'min_compare_len should be a positive integer')
+
+    if lookup_index.__len__() < 2 or lookup_index[0] > lookup_index[1]:
+        raise ValueError(
+            'lookup index should at least have 2 integer items, first specifying the start and second specifying the last indices')
+
+    if get_lookup_fn is None or type(get_lookup_fn) is not FunctionType:
+        raise TypeError(
+            'get_lookup_fn cannot be none and should be a function with 3 arguments')
```

Comment:
Do not worry about type checking. If user will give garbage, they will get garbage.